core/src/impl/Kokkos_Default_Graph_Impl.hpp

```diff
@@ -141,16 +141,29 @@ struct GraphImpl : private ExecutionSpaceInstanceStorage<ExecutionSpace> {
     m_has_been_instantiated = true;
   }
 
-  // FIXME The execution space instance should be passed to the nodes.
-  void submit(const ExecutionSpace& /* exec */) {
+  void submit(const ExecutionSpace& exec) {
     if (!m_has_been_instantiated) instantiate();
     // This reset is gross, but for the purposes of our simple host
     // implementation...
     for (auto& sink : m_sinks) {
       sink->reset_has_executed();
     }
+
+    // We don't know where the nodes will execute, so we need to fence the given
+    // execution space instance before proceeding. This is the simplest way
+    // of guaranteeing that the kernels in the graph are correctly "enqueued".
+    exec.fence(
+        "Kokkos::DefaultGraph::submit: fencing before launching graph nodes");
```

Comment:
This won't be sufficient for thread-safety.

Reply:
Could you explain what use case you have in mind ?

Here is mine.

When a user calls `graph.submit(exec_A)`, the expectation is that any node of the graph might start once all the work on `exec_A` is consumed.

Reply:
The thread-safety guarantee we have been working towards is that kernels committed to the same execution space instance don't run concurrently. So `graph.execute(exec_space)` should make sure that there are no kernels running concurrently to the graph execution on the same execution space instance.

I'm not saying we need to fix that here. We have to define thread-safety a little better in the documentation anyway.

Reply:
I guess that guaranteeing that there will never be 2 nodes that hold the same execution space instance that run concurrently would:
- either increase the complexity of the defaulted graph implementation (you would probably need to scan the graph and make smart routines here and there)
- or ensure that we run nodes one by one (sequentially), killing any chance that the defaulted graph would run anything asynchronously

So yes, I'd suggest we keep things as is, but keep your remark in mind :smile: